Dashboard app deploy icon

charlesfleche opened the pull request:
From the Dashboard, to reach the Deploy page of an application one has to first click on the app name, wait for the page to load, then click on Deploy. This PR adds an icon linked to the deploy page next to an application name. To avoid visual clutter this icon is only visible when the mouse hover over an application name.

![dashboard_app_deploy_icon](https://cloud.githubusercontent.com/assets/2563098/5626123/4358b1fe-95b5-11e4-9ef3-71ef1812b228.png)

These patches add a new apps_paginated_tile.html template (used by a new appsPaginatedTileDirective dashboard directive). A new TileType.APPS_PAGINATE now specializes the AppsPaginatedContext model.

Possible caveats :
- dashboardDirectives.appsPaginatedTileDirective and apps_paginated_tile.html come from brutals copy / paste. Some code refactoring may be necessary if the number of specializations starts to grow
- applications are not displayed using the default bootstap list-group because a link markup must not have interactive content (like another link) descendant. See http://stackoverflow.com/questions/9882916/are-you-allowed-to-nest-a-link-inside-of-a-link and http://www.w3.org/html/wg/drafts/html/master/single-page.html#the-a-element. In the standard bootstrap list-group an entire item can be an link markup : hovering the mouse over a line puts it in "link / hand" mode. But adding a clickable icon would violate the "no nested link" rule. In this PR, one has to click on the app name. Still, hovering in the item white space activates the grey background
- app and dashboard pages have slightly different icons as different versions of Awesome fonts seems to be used for these pages

charlesfleche committed:
Added the apps_paginated_tile dashboard ng_partial.

This partial displays applications' names along with an icon linked to the application 'Deploy page'. The icon is only visible when the mouse hover over the application line.

charlesfleche committed:
Removed debug comments

charlesfleche committed:
Added dashboard directive tileAppsPaginate

charlesfleche committed:
Changed AppsPaginatedContext.tile_type to a new type TileType.APPS_PAGINATE. Added key release_manager_url to paginated items.

dimagimon commented:
:+1: No lint errors found.

czue commented:
nice!

biyeun commented on corehq/apps/dashboard/static/dashboard/ng_partials/apps_paginated_tile.html:
```diff
@@ -0,0 +1,44 @@
+            <div class="list-group" ng-show="showItemList()">
+                <div class="list-group-item" data-ng-repeat="item in paginatedItems" >
+                   <a href="{{ item.url }}">{{ item.name }}</a>
+                   <a class="icon" href="{{ item.release_manager_url }}"><i class="icon-alternate-download fa fa-download"></i></a>
```
i wouldn't call this class `icon`, as that has historic meaning from Bootstrap 2 styles and will certainly cause confusion. Perhaps add a modifier like `secondary-link`. Also `icon-alternate-download` should not be used as the icon class, as we are not using glyphicons, but fontawesome

charlesfleche replied:
That makes sense, I'll fix that.

charlesfleche replied:
icon has been renamed  in a6e37e0
Reference to glyphicons has been deleted in 1673fca

biyeun commented on corehq/apps/dashboard/static/dashboard/js/hq_dashboard.ng.js:
```diff
@@ -180,6 +180,20 @@
         };
     };
 
+    dashboardDirectives.appsPaginatedTileDirective = function (dashboardConfig) {
```
can you think of a way to do this without adding a specific directive for applications? Making a secondary link option generic could potentially be useful for other paginated lists. Allowing the icon class to be specified for the secondary link would help with this.

biyeun commented:
@charlesfleche this is a great idea! just want to make sure this is implemented in a more generic way so that it can be useful for future tiles. :-)

charlesfleche committed:
Renamed dashboard class icon to list-group-item-icon

charlesfleche committed:
Removed dashboard reference to icon-alternate-download as glyphicons is not used in CommCareHQ

dimagimon commented:
:+1: No lint errors found.

charlesfleche commented:
Sure. I'll try to push a more generic version ASAP.

charlesfleche committed:
Refactored dashboard paginated view

models.py
- removed TileType.APP_PAGINATE
- AppsPaginatedContext uses _fmt_item
- Added two parameters to _fmt_item : secondary_url and secondary_url_icon

hq_dashboard.ng.js
- removed dashboardDirectives.appsPaginatedTileDirective

apps_paginated_tile.html
- removed this template

paginated_tile.html
- uses new item attributes secondary_url and secondary_url_icon
- output a <div><a /><a /></div> if a secondary url is defined
- output a simple <a> if no secondary url is defined

dimagimon commented on corehq/apps/dashboard/models.py:
```diff
@@ -307,13 +321,25 @@ def applications(self):
-            'name': a['key'][1],
-            'url': _get_app_url(a),
-        } for a in apps]
+        
```
W293 blank line contains whitespace

dimagimon commented on corehq/apps/dashboard/models.py:
```diff
@@ -307,13 +321,25 @@ def applications(self):
+        
+        return [self._fmt_item(_get_app_name(a),
+                               _get_app_url(a),
+                               None, # description
```
E261 at least two spaces before inline comment

dimagimon commented on corehq/apps/dashboard/models.py:
```diff
@@ -307,13 +321,25 @@ def applications(self):
+        return [self._fmt_item(_get_app_name(a),
+                               _get_app_url(a),
+                               None, # description
+                               None, # full_name
```
E261 at least two spaces before inline comment

charlesfleche commented:
bd96784 was supposed to make this list display more generic… In this version there only one pagination template. It receives an item. Depending on the values of item.secondary_url and item.secondary_item_icon, an icon is added (or not) at the end of each lilne.

It almost work, until I found that angularjs doesn't allow condition in its template. It what I tried to do here, with a {% if item.secondary_url and item.secondaray_url_icon %}{% else %}{% endif %} block.

I'm very neophyte when it comes to web development and right now I wonder what would the best way to continue this branch. It seems like paginated_tile.html should first be used in angularjs, then another template within paginated_tile.html should be selected depending on the secondary_url* values.

What do you guys think ?

charlesfleche committed:
Fixing dimagimon warnings

dimagimon commented:
:+1: No lint errors found.

charlesfleche commented:
Error timed out while getting http://files.dimagi.com/pygooglechart-0.3.0.tar.gz

biyeun commented on corehq/apps/dashboard/static/dashboard/ng_partials/paginated_tile.html:
```diff
@@ -14,17 +14,33 @@
-                   title="">
-                    {{ item.name }}
-                </a>
+                {% if item.secondary_url and item.secondary_url_icon %}
```
Be careful with this, as angular templates may share similar syntax with django templates but ultimately they are two different things. See the `ng-if` directive for what you'd like to do here: https://docs.angularjs.org/api/ng/directive/ngIf

charlesfleche committed:
Uses ngIf to switch between a simple anchor or an anchor + icon, depending on item.secondary_url

dimagimon commented:
:+1: No lint errors found.

biyeun commented:
looks good. waiting for build

czue commented:
thanks @charlesfleche !
